auth/src/main/java/com/redhat/lightblue/rest/auth/ldap/LdapFindUserByUidCommand.java

```diff
@@ -1,62 +1,132 @@
+           //propagate the exception
+            throw e;
+        } catch (LDAPUserNotFoundException | LDAPMutipleUserFoundException e) {
+            // Return null in case the User not found or multiple Users were found (which is inconsistent)
 
-        NamingEnumeration<SearchResult> results = ldapContext.search(ldapSearchBase, searchFilter, searchControls);
+            if(e instanceof LDAPUserNotFoundException)
+                LOGGER.info("No result found roles for user: " + cacheKey.uid, e);
+            else{
+                LOGGER.error("Multiples users found and only one was expected for user: " + cacheKey.uid, e);
+            }
 
-        SearchResult searchResult = null;
-        if (results.hasMoreElements()) {
-            searchResult = results.nextElement();
+            searchResult = LDAPCache.getLDAPCacheSession().getIfPresent(cacheKey);
+            if (searchResult != null) {
+                // if (not found on the server OR server state is inconsistent ) and cache hold the old value, evict the entry
+                LDAPCache.invalidateKey(cacheKey);
+            }
+        }
+        LDAPCache.getLDAPCacheSession().put(cacheKey, searchResult);
+
+        return searchResult;
+    }
 
-            //make sure there is not another item available, there should be only 1 match
-            if (results.hasMoreElements()) {
-                LOGGER.error("Matched multiple users for the accountName: " + uid);
+    /*
+        This methods is executed for all types of failure such as run() failure, timeout, thread pool or semaphore rejection, and circuit-breaker short-circuiting
+     */
+    @Override
+    protected SearchResult getFallback() {
+        LOGGER.info("Error during the execution of the command. Falling back to the cache");
+        return new FallbackViaLDAPServerProblemCommand(cacheKey, getFailedExecutionException()).execute();
+
+    }
+
+
+    /*
+        Use the cache in case tje LDAP Server was not available and also to we have metrics around the fallback
+     */
+    private static class FallbackViaLDAPServerProblemCommand extends HystrixCommand<SearchResult> {
+        private static final Logger LOGGER = Logger.getLogger(FallbackViaLDAPServerProblemCommand.class);
+
+        private final LDAPCacheKey cacheKey;
+        private final Throwable failedExecutionThrowable;
+
+        public FallbackViaLDAPServerProblemCommand(LDAPCacheKey cacheKey, Throwable failedExecutionThrowable) {
+            super(HystrixCommand.Setter.withGroupKey(HystrixCommandGroupKey.Factory.asKey(GROUPKEY)).
+                    andCommandKey(HystrixCommandKey.Factory.asKey(GROUPKEY + ":" + FallbackViaLDAPServerProblemCommand.class.getSimpleName())));
+            this.cacheKey = cacheKey;
+            this.failedExecutionThrowable = failedExecutionThrowable;
+        }
+
+        @Override
+        protected SearchResult run() throws Exception {
```

Comment:
It relies on the cache when It couldn't connect to the LDAP server (which I think it is the use case). I didn't understand why would all the request would fail as you mentioned, unless you are talking about Error from the JVM. About the Exception propagation,  roughly it is doing that , the other propagations that you can see in that method are for cases that we don't handle but Hystrix & Java best practices suggest to us to propagate or even exist the process